### PR TITLE
dead-letter-queue

### PR DESCRIPTION
**Pull request by prithvi081099**

### CODE REVIEW

File: gateway/modules/eventing/helpers.go [hunk 1]
```diff
@@ -179,6 +179,22 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
...
+		Payload: map[string]interface{}{
+			"id":   eventDoc.ID,
+			"body": eventDoc.Payload,
+		},
+	}
+
+	_, err := m.QueueEvent(context.Background(), m.project, "", &req)
+	if err != nil {
+		logrus.Errorf("error=%v\n", err)
+	}
```
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:194]: No need to check for error, directly return

File: gateway/modules/eventing/handle_staged.go [hunk 2]
```diff
@@ -132,7 +132,9 @@ func (m *Module) processStagedEvent(eventDoc *model.EventDocument) {
 		// Reaching here means the event was successfully processed. Let's simply return
 		return
 	}
-
+	if err := m.generateDLQEventRequest(eventDoc); err != nil {
```
Comment (sharadregoti) [inline gateway/modules/eventing/handle_staged.go:135]: pass context available above in proccessStageEvent Function

File: gateway/modules/eventing/helpers.go [hunk 3]
```diff
@@ -179,6 +179,22 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
 	}
 }
 
+func (m *Module) generateDLQEventRequest(eventDoc *model.EventDocument) error {
```
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:182]: take the context as first param

File: gateway/modules/eventing/helpers.go [hunk 4]
```diff
@@ -179,6 +179,22 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
...
+	req := model.QueueEventRequest{
+		Type: "dlq_event_trigger_" + eventDoc.RuleName,
```
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:184]: Use fmt.sprintf() & store the "dlq_event_trigger_" as a contant in utils/ constants.go file. name it as DLQEventTriggerSuffix

File: gateway/modules/eventing/helpers.go [hunk 5]
```diff
@@ -179,6 +179,22 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
...
+		Payload: map[string]interface{}{
+			"id":   eventDoc.ID,
+			"body": eventDoc.Payload,
+		},
+	}
+
+	_, err := m.QueueEvent(context.Background(), m.project, "", &req)
```
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:191]: use the ctx passed above. & use m.adminman > getinternalaccesstoken function to generate token

File: gateway/modules/eventing/helpers.go (hunk 3 above)
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:182]:
```suggestion
func (m *Module) triggerDLQEvent(eventDoc *model.EventDocument) error {
```

File: gateway/modules/eventing/handle_staged.go [hunk 6]
```diff
@@ -132,7 +132,9 @@ func (m *Module) processStagedEvent(eventDoc *model.EventDocument) {
...
+		logrus.Errorf(err.Error())
```
Comment (sharadregoti) [inline gateway/modules/eventing/handle_staged.go:136]: Use utils.LogErrror() with message "Couldn't create DLQ event for event id %v" eventdoc.id

File: gateway/model/crud.go [hunk 7]
```diff
@@ -65,3 +65,10 @@ type AllRequest struct {
 type BatchRequest struct {
 	Requests []*AllRequest `json:"reqs"`
 }
+
+// DLQRequest is request body for dlq
+type DLQRequest struct {
+	ID   string      `json:"id"`
+	Type string      `json:"type"`
+	Body interface{} `json:"body"`
+}
```
Comment (sharadregoti) [inline gateway/model/crud.go:74]: I think this is not required, as you are not using it anywhere

File: gateway/modules/eventing/helpers.go [hunk 8]
```diff
@@ -179,6 +179,22 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
 	}
 }
 
+func (m *Module) triggerDLQEvent(ctx context.Context, eventDoc *model.EventDocument) error {
+	req := model.QueueEventRequest{
+		Type: fmt.Sprintf("%s%s", utils.DLQEventTriggerSuffix, eventDoc.RuleName),
+		Payload: map[string]interface{}{
+			"id":   eventDoc.ID,
```
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:186]: also add the type of event
"type":eventdoc.Type

File: gateway/modules/eventing/helpers.go [hunk 9]
```diff
@@ -179,6 +179,23 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
...
+			"type": eventDoc.Type,
```
Comment (YourTechBud) [inline gateway/modules/eventing/helpers.go:187]:
```suggestion
			"event_type": eventDoc.Type,
```

File: gateway/modules/eventing/helpers.go [hunk 10]
```diff
@@ -179,6 +179,23 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
...
+			"body": eventDoc.Payload,
```
Comment (YourTechBud) [inline gateway/modules/eventing/helpers.go:188]:
```suggestion
			"event_payload": eventDoc.Payload,
```

Comment (YourTechBud) [reply gateway/modules/eventing/helpers.go:188]: Also, make sure that the `eventDoc.Payload` is an objectand not `[]byte` or `string`.

File: gateway/modules/eventing/helpers.go [hunk 11]
```diff
@@ -179,6 +179,23 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
 	}
 }
 
+func (m *Module) triggerDLQEvent(ctx context.Context, eventDoc *model.EventDocument) error {
+	req := model.QueueEventRequest{
+		Type: fmt.Sprintf("%s%s", utils.DLQEventTriggerSuffix, eventDoc.RuleName),
+		Payload: map[string]interface{}{
+			"id":   eventDoc.ID,
```
Comment (YourTechBud) [inline gateway/modules/eventing/helpers.go:186]: Add the following fields to the payload
- prevent event ts
- previous trigger name

File: gateway/utils/constants.go [hunk 12]
```diff
@@ -7,6 +7,9 @@ import (
 // BuildVersion is the current version of Space Cloud
 const BuildVersion = "0.18.0"
 
+//DLQEventTriggerSuffix used as suffix for DLQ event trigger
+const DLQEventTriggerSuffix = "dlq_event_trigger_"
```
Comment (YourTechBud) [inline gateway/utils/constants.go:11]:
```suggestion
const DLQEventTriggerPrefix = "dlq_"
```

File: gateway/modules/eventing/helpers.go [hunk 13]
```diff
@@ -179,6 +179,23 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
...
+			"type": eventDoc.Type,
+			"body": eventDoc.Payload,
+		},
+	}
+	token, err := m.adminMan.GetInternalAccessToken()
+	if err != nil {
+		return err
+	}
+	_, err = m.QueueEvent(ctx, m.project, token, &req)
```
Comment (YourTechBud) [inline gateway/modules/eventing/helpers.go:195]: Can we skip the schema validation and security rules step? That could create a problem. Use an internal method

File: gateway/modules/eventing/helpers.go [hunk 14]
```diff
@@ -179,6 +180,33 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
 	}
 }
 
+func (m *Module) triggerDLQEvent(ctx context.Context, eventDoc *model.EventDocument) error {
+	if reflect.TypeOf(eventDoc.Payload).String() == "string" || reflect.TypeOf(eventDoc.Payload) == reflect.TypeOf([]byte(nil)) {
+		var doc interface{}
+		_ = json.Unmarshal([]byte(eventDoc.Payload.(string)), &doc)
+		eventDoc.Payload = doc
+	}
```
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:188]: @YourTechBud is this required ?

File: gateway/modules/eventing/helpers.go [hunk 15]
```diff
@@ -179,6 +180,33 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
...
+		},
+	}
+
+	if err := m.batchRequests(ctx, []*model.QueueEventRequest{req}, m.generateBatchID()); err != nil {
+		logrus.Errorf("error queueing dlq event in eventing unable to batch requests - %s", err.Error())
+		return err
+	}
+
+	m.metricHook(m.project, req.Type)
+	//_, err = m.QueueEvent(ctx, m.project, token, &req)
```
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:206]: remove this comment

File: gateway/modules/eventing/helpers.go (hunk 14 above)
Comment (YourTechBud) [reply gateway/modules/eventing/helpers.go:188]: Wait what. No this looks wrong

File: gateway/modules/eventing/helpers.go [hunk 16]
```diff
@@ -179,6 +180,33 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
 	}
 }
 
+func (m *Module) triggerDLQEvent(ctx context.Context, eventDoc *model.EventDocument) error {
+	switch eventDoc.Payload.(type) {
+	case map[string]interface{}:
+	default:
+		logrus.Errorf("Payload in given event is of type %v wanted type map[string]interface{}", reflect.TypeOf(eventDoc.Payload))
```
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:187]: use utils.Log error
```suggestion
		logrus.Errorf("Payload in given event is of type %v wanted type object", reflect.TypeOf(eventDoc.Payload))
```

File: gateway/modules/eventing/helpers.go [hunk 17]
```diff
@@ -179,6 +180,33 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
...
+			"event_id":        eventDoc.ID,
+			"event_type":      eventDoc.Type,
+			"event_payload":   eventDoc.Payload,
+			"event_timestamp": eventDoc.Timestamp,
+			"event_name":      eventDoc.RuleName,
+		},
+	}
+
+	if err := m.batchRequests(ctx, []*model.QueueEventRequest{req}, m.generateBatchID()); err != nil {
+		logrus.Errorf("error queueing dlq event in eventing unable to batch requests - %s", err.Error())
```
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:202]: use utils.log error

File: gateway/modules/eventing/helpers.go (hunk 16 above)
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go:187]: try using if statement.
if reflect.typeof(ev.payload).kind() != reflect.Map {
throw error
}

**Comment by sharadregoti** [comment]
@prithvi081099 resolve the reviews ASAP

File: gateway/modules/eventing/helpers.go [hunk 18]
```diff
@@ -179,6 +180,31 @@ func (m *Module) generateProcessedEventRequest(eventID string) *model.UpdateRequ
 	}
 }
 
+func (m *Module) triggerDLQEvent(ctx context.Context, eventDoc *model.EventDocument) error {
+	if reflect.TypeOf(eventDoc.Payload).Kind() != reflect.Map {
+		_ = utils.LogError(fmt.Sprintf("Payload in given event is of type %v wanted type object", reflect.TypeOf(eventDoc.Payload)), "eventing", "triggerDLQEvent", nil)
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply gateway/modules/eventing/helpers.go]: so what do you suggest, golang type switch?

**Pull request merged**